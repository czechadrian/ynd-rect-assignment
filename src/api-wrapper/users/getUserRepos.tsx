import { TGetUserReposResponse } from '../types';
import { apiGet, createUrl } from '../apiCall';

export const getAllUserRepos = async (
  user: string | Array<string>
): Promise<TGetUserReposResponse> => {
  const url = createUrl(`https://api.github.com/users/${user}/repos`);

  return apiGet<TGetUserReposResponse>(url);
};
