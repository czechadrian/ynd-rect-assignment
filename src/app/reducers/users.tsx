import { createReducer } from 'typesafe-actions';
import produce from 'immer';

import {
  fetchUsersInitAction,
  fetchUsersSuccessAction,
  fetchUsersFailureAction,
  TFetchUsersActions
} from '../actions/users';
import { TFetchingStatus } from '../helpers';

type TUsers = {
  name: string;
  fetchingStatus: TFetchingStatus;
};

export const initialState: TUsers = {
  name: '',
  fetchingStatus: TFetchingStatus.Defined
};

export const usersReducer = createReducer<TUsers, TFetchUsersActions>(
  initialState
)
  .handleAction(fetchUsersInitAction, (state) =>
    produce(state, (draftState) => {
      draftState.fetchingStatus = TFetchingStatus.Initial;
    })
  )
  .handleAction(fetchUsersSuccessAction, (state, action) =>
    produce(state, (draftState) => {
      draftState.fetchingStatus = TFetchingStatus.Success;
      draftState.name = action.payload;
    })
  )
  .handleAction(fetchUsersFailureAction, (state) =>
    produce(state, (draftState) => {
      draftState.fetchingStatus = TFetchingStatus.Failure;
    })
  );
