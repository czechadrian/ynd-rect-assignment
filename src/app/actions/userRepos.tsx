import { ThunkAction } from 'redux-thunk';
import { Action, ActionType, createAction } from 'typesafe-actions';

import { TRootState } from '../reducers';
import { getAllUserRepos } from '../../api-wrapper/users/getUserRepos';
import { TGetUserReposResponse } from '../../api-wrapper/types';

export const fetchUserReposInitAction = createAction('user/GET_REPOS_INIT')();
export const fetchUserReposSuccessAction = createAction(
  'user/GET_REPOS_SUCCESS'
)<TGetUserReposResponse>();
export const fetchUserReposFailureAction = createAction(
  'user/GET_REPOS_FAILURE'
)();

export type TFetchUserReposInitAction = ActionType<
  typeof fetchUserReposInitAction
>;
export type TFetchUserReposSuccessAction = ActionType<
  typeof fetchUserReposSuccessAction
>;
export type TFetchUserReposFailureAction = ActionType<
  typeof fetchUserReposFailureAction
>;

export type TFetchUserReposActions =
  | TFetchUserReposInitAction
  | TFetchUserReposFailureAction
  | TFetchUserReposSuccessAction;

export type TUserReposThunkActionType = (
  search: string | Array<string>
) => ThunkAction<void, TRootState, null, TFetchUserReposActions>;

export const getUserRepos: TUserReposThunkActionType =
  (search: string | Array<string>) =>
  async (dispatch: <T>(action: Action | Promise<Action>) => T) => {
    dispatch(fetchUserReposInitAction());
    return getAllUserRepos(search)
      .then((payload) => {
        return dispatch(fetchUserReposSuccessAction(payload));
      })
      .catch(() => dispatch(fetchUserReposFailureAction()));
  };
