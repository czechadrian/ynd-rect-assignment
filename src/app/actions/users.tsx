import { ThunkAction } from 'redux-thunk';
import { Action, ActionType, createAction } from 'typesafe-actions';

import { TRootState } from '../reducers';

export const fetchUsersInitAction = createAction('users/FETCH_INIT')();
export const fetchUsersSuccessAction = createAction(
  'users/FETCH_SUCCESS'
)<string>();
export const fetchUsersFailureAction = createAction('users/FETCH_FAILURE')();

export type TFetchUsersInitAction = ActionType<typeof fetchUsersInitAction>;
export type TFetchUsersSuccessAction = ActionType<
  typeof fetchUsersSuccessAction
>;
export type TFetchUsersFailureAction = ActionType<
  typeof fetchUsersFailureAction
>;

export type TFetchUsersActions =
  | TFetchUsersInitAction
  | TFetchUsersFailureAction
  | TFetchUsersSuccessAction;

export type TUsersThunkActionType = ThunkAction<
  void,
  TRootState,
  null,
  TFetchUsersActions
>;

export const getUsers: TUsersThunkActionType =
  () => async (dispatch: <T>(action: Action | Promise<Action>) => T) => {
    dispatch(fetchUsersInitAction());
    // return getUsersAPI()
    //     .then((payload) => {
    //         return dispatch(fetchUsersSuccessAction(payload));
    //     })
    //     .catch(() => dispatch(fetchUsersFailureAction()));
  };
