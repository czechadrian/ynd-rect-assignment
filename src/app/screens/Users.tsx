import React, { useEffect } from 'react';
import {Empty, notification, Spin} from 'antd';
import _ from "lodash";
import { useDispatch, useSelector } from 'react-redux';

import { TButtonSize, TButtonTypes, TFetchingStatus } from '../helpers';
import { getUsers } from '../actions/users';
import {
  getUsersFetchingStatusSelector,
  getUsersSelector
} from '../../selectors';
import UserList from '../components/UserList';
import { displayShowingResultsLabel } from '../helpers/utils';
import {
  ButtonStyled,
  InputStyled,
  Layout,
  ShowingResultsStyled
} from './styled';
import {SpinWrapper} from "../components/styled";

const Users: React.FC = () => {
  const dispatch = useDispatch();
  const users = useSelector(getUsersSelector);
  const usersFetchingStatus = useSelector(getUsersFetchingStatusSelector);
  const [search, setSearch] = React.useState('');

  const searchHandler = (search: string) => {
    setSearch(search);
  };

  const displayData = () =>
    users.length === 0 ? <Empty /> : <UserList users={users} />;

  useEffect(() => {
    if (_.isEqual(usersFetchingStatus, TFetchingStatus.Failure))
      notification.error({
        message: 'Error',
        description: 'Fetching users data failed.'
      });
    else if (_.isEqual(usersFetchingStatus, TFetchingStatus.Success))
      notification.success({
        message: 'Success',
        description: 'Fetching users data succeeded.'
      });
  }, [usersFetchingStatus]);

  return (
    <Layout>
      <InputStyled
        size={TButtonSize.LARGE}
        placeholder="Enter username"
        onChange={(val) => searchHandler(val.target.value)}
      />
      <ButtonStyled
        size={TButtonSize.LARGE}
        type={TButtonTypes.PRIMARY}
        onClick={() => dispatch(getUsers(search))}
      >
        Search
      </ButtonStyled>
      <ShowingResultsStyled>
        {displayShowingResultsLabel(usersFetchingStatus, search)}
      </ShowingResultsStyled>
      {_.isEqual(usersFetchingStatus, TFetchingStatus.Success) && displayData()}
      {_.isEqual(usersFetchingStatus, TFetchingStatus.Initial) && (
          <SpinWrapper>
            <Spin size="large" />
          </SpinWrapper>)}
    </Layout>
  );
};

export default Users;
