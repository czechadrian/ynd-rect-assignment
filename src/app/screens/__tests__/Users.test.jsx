import React from "react";
import { useDispatch, useSelector } from "react-redux";
import "jest-styled-components";
import renderer from "react-test-renderer";
import Users from "../Users";
import {getUsers} from "../../actions/users";

jest.mock("../../actions/users", () => {
    return {
        getUsers: jest.fn(),
    };
});

jest.mock("react-redux", () => {
    return {
        useSelector: jest.fn(),
        useDispatch: jest.fn(),
    };
});

describe("Users", () => {
    const dispathMock = jest.fn();
    beforeEach(() => {
        useSelector.mockImplementation(() => [{login:"user login"}]);
        useDispatch.mockReturnValue(dispathMock);

        getUsers.mockReturnValue(() => jest.fn());

        jest.clearAllMocks();
    });
    test("selected comments component test", () => {
        const tree = renderer.create(<Users />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});