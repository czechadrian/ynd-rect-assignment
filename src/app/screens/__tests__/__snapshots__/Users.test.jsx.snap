// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users selected comments component test 1`] = `
.c0 {
  width: 480px;
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: white;
}

.c1 {
  width: 100%;
  background-color: #e8e8e8;
  margin-bottom: 1rem;
  color: #707070;
}

.c1::placeholder {
  color: #707070;
}

.c2 {
  width: 100%;
}

.c3 {
  width: 100%;
  padding: 0.5rem;
  font-size: 1.25rem;
}

@media (max-width:800px) {
  .c0 {
    height: 700px;
  }
}

@media (min-width:801px) {
  .c0 {
    height: 95%;
  }
}

<div
  className="c0"
>
  <input
    className="ant-input ant-input-lg c1"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="Enter username"
    type="text"
    value=""
  />
  <button
    className="ant-btn ant-btn-primary ant-btn-lg c2"
    onClick={[Function]}
    type="button"
  >
    <span>
      Search
    </span>
  </button>
  <span
    className="c3"
  />
</div>
`;
